staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go

```diff
@@ -149,290 +169,196 @@ func PatchResource(r rest.Patcher, scope RequestScope, admit admission.Interface
-			}
+	trace *utiltrace.Trace
 
-			versionedObjToUpdate, err := creater.New(kind)
-			if err != nil {
-				return nil, err
-			}
-			if err := applyPatchToObject(codec, defaulter, currentObjMap, originalPatchMap, versionedObjToUpdate, versionedObj); err != nil {
-				return nil, err
-			}
-			// Convert the object back to unversioned.
-			gvk := kind.GroupKind().WithVersion(runtime.APIVersionInternal)
-			objToUpdate, err := unsafeConvertor.ConvertToVersion(versionedObjToUpdate, gvk.GroupVersion())
-			if err != nil {
-				return nil, err
-			}
+	// Set at construction time and immutable
+	namespace         string
+	updatedObjectInfo rest.UpdatedObjectInfo
+	mechanism         patchMechanism
 
-			return objToUpdate, nil
-		}
-	}
+	// Modified each iteration
+	iterationCount  int
+	lastConflictErr error
 
-	// applyAdmission is called every time GuaranteedUpdate asks for the updated object,
-	// and is given the currently persisted object and the patched object as input.
-	applyAdmission := func(ctx request.Context, patchedObject runtime.Object, currentObject runtime.Object) (runtime.Object, error) {
-		trace.Step("About to check admission control")
-		return patchedObject, updateMutation(patchedObject, currentObject)
-	}
-	updatedObjectInfo := rest.DefaultUpdatedObjectInfo(nil, applyPatch, applyAdmission)
+	// Set on first iteration
+	originalResourceVersion string
+}
 
-	return finishRequest(timeout, func() (runtime.Object, error) {
-		updateObject, _, updateErr := patcher.Update(ctx, name, updatedObjectInfo, createValidation, updateValidation)
-		for i := 0; i < MaxRetryWhenPatchConflicts && (errors.IsConflict(updateErr)); i++ {
-			lastConflictErr = updateErr
-			updateObject, _, updateErr = patcher.Update(ctx, name, updatedObjectInfo, createValidation, updateValidation)
-		}
-		return updateObject, updateErr
-	})
+func (p *patcher) toUnversioned(versionedObj runtime.Object) (runtime.Object, error) {
+	gvk := p.kind.GroupKind().WithVersion(runtime.APIVersionInternal)
+	return p.unsafeConvertor.ConvertToVersion(versionedObj, gvk.GroupVersion())
 }
 
-// patchObjectJSON patches the <originalObject> with <patchJS> and stores
-// the result in <objToUpdate>.
-// Currently it also returns the original and patched objects serialized to
-// JSONs (this may not be needed once we can apply patches at the
-// map[string]interface{} level).
-func patchObjectJSON(
-	patchType types.PatchType,
-	codec runtime.Codec,
-	originalObject runtime.Object,
-	patchJS []byte,
-	objToUpdate runtime.Object,
-	versionedObj runtime.Object,
-) (originalObjJS []byte, patchedObjJS []byte, retErr error) {
-	js, err := runtime.Encode(codec, originalObject)
+type patchMechanism interface {
+	firstPatchAttempt(currentObject runtime.Object, currentResourceVersion string) (runtime.Object, error)
+	subsequentPatchAttempt(currentObject runtime.Object, currentResourceVersion string) (runtime.Object, error)
+}
+
+type jsonPatcher struct {
+	*patcher
```

Comment:
inlining the patcher in the mechanism makes it hard to tell which fields of jsonPatcher are safe to mutate during the operation and which are supposed to be immutable

Reply:
That criticism applies even harder to the original code, not sure if I will make further changes or not.

Reply:
well, the mutations are hidden in methods now, which makes them harder to trace... my preference would be for more separation of the immutable values (even just not inlining to make it obvious when you were mutating something that didn't belong to the mechanism)